### PR TITLE
Reduce work done in operator overload syntax classifier.

### PR DESCRIPTION
**Pull request by ToddGrun**
This shows up as 15% of CPU in the code analysis process in a profile provided by DavidW where semantic token requests are performing poorly.

Roslyn classification walks down the tree, finding AssignmentExpressionSyntax nodes and performing semantic binding on the nodes (expensive).

Two optimizations to short-circuit before performing the binding: 1) Per Cyrus, this operation doesn't need to occur on simple assignments (the most common case of AssignmentExpressionSyntax nodes) 2) Check whether the requested range intersects the location where the classification would actually be reported.

### CODE REVIEW

**Comment by ToddGrun** [comment]
image

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.cs [hunk 1]
```diff
@@ -23,17 +23,23 @@ internal class OperatorOverloadSyntaxClassifier : AbstractSyntaxClassifier
...
+            TextSpan textSpan,
             SemanticModel semanticModel,
             ClassificationOptions options,
             SegmentedList<ClassifiedSpan> result,
             CancellationToken cancellationToken)
         {
-            var symbolInfo = semanticModel.GetSymbolInfo(syntax, cancellationToken);
-            if (symbolInfo.Symbol is IMethodSymbol methodSymbol
-                && methodSymbol.MethodKind == MethodKind.UserDefinedOperator)
+            if (syntax.IsKind(SyntaxKind.SimpleAssignmentExpression))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.cs:32]: doc.

File: src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.cs [hunk 2]
```diff
@@ -23,17 +23,23 @@ internal class OperatorOverloadSyntaxClassifier : AbstractSyntaxClassifier
...
-                if (!operatorSpan.IsEmpty)
+                return;
+            }
+
+            var operatorSpan = GetOperatorTokenSpan(syntax);
+            if (!operatorSpan.IsEmpty && operatorSpan.IntersectsWith(textSpan))
+            {
+                var symbolInfo = semanticModel.GetSymbolInfo(syntax, cancellationToken);
+                if (symbolInfo.Symbol is IMethodSymbol methodSymbol
+                    && methodSymbol.MethodKind == MethodKind.UserDefinedOperator)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.cs:42]:
```suggestion
                if (symbolInfo.Symbol is IMethodSymbol { MethodKind: MethodKind.UserDefinedOperator })
```

File: src/Workspaces/VisualBasic/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.vb [hunk 3]
```diff
@@ -20,20 +21,35 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Classification.Classifiers
...
+        Private Shared Function GetOperatorTokenSpan(syntax As SyntaxNode) As TextSpan
+            If TypeOf syntax Is BinaryExpressionSyntax Then
+                Return DirectCast(syntax, BinaryExpressionSyntax).OperatorToken.Span
+            ElseIf TypeOf syntax Is UnaryExpressionSyntax Then
+                Return DirectCast(syntax, UnaryExpressionSyntax).OperatorToken.Span
+            End If
+
+            Return Nothing
+        End Function
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.vb:53]:
```suggestion
```

**Review by CyrusNajmabadi — APPROVED** [review]

**Review by genlu — APPROVED** [review]

File: src/Workspaces/VisualBasic/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.vb [hunk 4]
```diff
@@ -26,18 +26,22 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Classification.Classifiers
...
                 Return
             End If
 
+            ' Short-circuit operators whose span doesn't intersect the requested span.
             Dim operatorSpan = GetOperatorTokenSpan(syntax)
-            If (Not operatorSpan.IsEmpty AndAlso operatorSpan.IntersectsWith(textSpan)) Then
-                Dim symbolInfo = semanticModel.GetSymbolInfo(syntax, cancellationToken)
-                If TypeOf symbolInfo.Symbol Is IMethodSymbol AndAlso
-                    DirectCast(symbolInfo.Symbol, IMethodSymbol).MethodKind = MethodKind.UserDefinedOperator Then
+            If (operatorSpan.IsEmpty OrElse Not operatorSpan.IntersectsWith(textSpan)) Then
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.vb:36]:
```suggestion
            If operatorSpan.IsEmpty OrElse Not operatorSpan.IntersectsWith(textSpan) Then
```

Comment (ToddGrun) [reply src/Workspaces/VisualBasic/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.vb:36]: C# habits die hard

Comment (ToddGrun) [reply src/Workspaces/VisualBasic/Portable/Classification/SyntaxClassification/OperatorOverloadSyntaxClassifier.vb:36]: Bah, it doesn't look like this change was pushed before merged. Will create a small PR with just that.

**Pull request merged**